controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java

```diff
@@ -25,19 +29,32 @@
     protected void maintain() {
 
         for (Application application : controller().applications().asList()) {
-            try (Lock lock = controller().applications().lock(application.id())) {
-                for (Deployment deployment : application.deployments().values()) {
-
+            for (Deployment deployment : application.deployments().values()) {
+                try {
                     MetricsService.DeploymentMetrics metrics = controller().metricsService()
                             .getDeploymentMetrics(application.id(), deployment.zone());
-
                     DeploymentMetrics appMetrics = new DeploymentMetrics(metrics.queriesPerSecond(), metrics.writesPerSecond(),
-                            metrics.documentCount(), metrics.queryLatencyMillis(), metrics.writeLatencyMillis());
-
-                    Application app = application.with(deployment.withMetrics(appMetrics));
-                    controller().applications().store(app, lock);
+                                                                         metrics.documentCount(), metrics.queryLatencyMillis(), metrics.writeLatencyMillis());
+
+                    // Avoid locking for a long time, due to slow YAMAS.
+                    try (Lock lock = controller().applications().lock(application.id())) {
+                        // Deployment (or application) may have changed (or be gone) now:
+                        controller().applications().get(application.id()).ifPresent(freshApplication -> {
```

Comment:
Imho this code would be cleaner by using 2 instances of
if (not present) continue;
instead of the nesting and 2 different ways of bailing out.